napari_cellseg3d/model_workers.py

```diff
@@ -289,11 +299,330 @@ def log_parameters(self):
+            + f"_{utils.get_date_time()}_"
+            + self.filetype
+        )
+
+        imwrite(instance_filepath, instance_labels)
+        self.log(
+            f"Instance segmentation results for image n°{image_id} have been saved as:"
+        )
+        self.log(os.path.split(instance_filepath)[1])
+        return instance_labels
+        # print(self.stats_to_csv)
+
+    def inference_on_list(self, inf_data, i, model, post_process_transforms):
+
+        self.log("-" * 10)
+        self.log(f"Inference started on image n°{i + 1}...")
+
+        inputs = inf_data["image"]
+        # print(inputs.shape)
+
+        out = self.model_output(inputs, model, post_process_transforms)
+        out = self.aniso_transform(out)
+
+        out = np.array(out).astype(np.float32)
+        out = np.squeeze(out)
+        to_instance = out  # avoid post processing since thresholding is done there anyway
+        image_id = i + 1
+        time = utils.get_date_time()
+        # print(time)
+
+        original_filename = os.path.basename(self.images_filepaths[i]).split(
+            "."
+        )[0]
+
+        # File output save name : original-name_model_date+time_number.filetype
+        file_path = (
+            self.results_path
+            + "/"
+            + f"Prediction_{image_id}_"
+            + original_filename
+            + "_"
+            + self.model_dict["name"]
+            + f"_{time}_"
+            + self.filetype
+        )
+
+        # print(filename)
+        imwrite(file_path, out)
+
+        self.log(f"\nFile n°{image_id} saved as :")
+        filename = os.path.split(file_path)[1]
+        self.log(filename)
+
+        #################
+        #################
+        #################
+        if self.instance_params["do_instance"]:
+            instance_labels = self.instance_seg(
+                to_instance, image_id, original_filename
+            )
+            data_dict = self.stats_csv(instance_labels)
+        else:
+            instance_labels = None
+            data_dict = None
+
+        original = np.array(inf_data["image"]).astype(np.float32)
+
+        self.log(f"Inference completed on layer")
+        result = {
+            "image_id": i + 1,
+            "original": original,
+            "instance_labels": instance_labels,
+            "object stats": data_dict,
+            "result": out,
+            "model_name": self.model_dict["name"],
+        }
+        return result
+
+    def stats_csv(self, instance_labels):
+        if self.stats_to_csv:
+
+            # try:
+
+            data_dict = volume_stats(
+                instance_labels
+            )  # TODO test with area mesh function
+            return data_dict
+
+        # except ValueError as e:
+        #     self.log(f"Error occurred during stats computing : {e}")
+        #     return None
+        else:
+            return None
+
+    def inference_on_layer(self, image, model, post_process_transforms):
```

Comment:
Seems to be a bunch of shared code with inference_on_list, could you call a new shared function from both of these ?

Reply:
Fixed in 8cf755c